demos/mqtt/mqtt_demo_mutual_auth/mqtt_demo_mutual_auth.c

```diff
@@ -0,0 +1,1520 @@
+        bytesSent = -1;
+    }
+
+    return bytesSent;
+}
+
+/*-----------------------------------------------------------*/
+
+static int32_t transportRecv( MQTTNetworkContext_t pSslContext,
+                              void * pBuffer,
+                              size_t bytesToRecv )
+{
+    int32_t bytesReceived = 0;
+    int pollStatus = -1, bytesAvailableToRead = 0;
+    struct pollfd fileDescriptor =
+    {
+        .events  = POLLIN | POLLPRI,
+        .revents = 0
+    };
+
+    /* Set the file descriptor for poll. */
+    fileDescriptor.fd = SSL_get_fd( pSslContext );
+
+    /* Check if there are any pending data available for read. */
+    bytesAvailableToRead = SSL_pending( pSslContext );
+
+    /* Poll only if there is no data available yet to read. */
+    if( bytesAvailableToRead <= 0 )
+    {
+        pollStatus = poll( &fileDescriptor, 1, TRANSPORT_SEND_RECV_TIMEOUT_MS );
+    }
+
+    /* SSL read of data. */
+    if( ( pollStatus > 0 ) || ( bytesAvailableToRead > 0 ) )
+    {
+        bytesReceived = ( int32_t ) SSL_read( pSslContext, pBuffer, bytesToRecv );
+    }
+    /* Poll timed out. */
+    else if( pollStatus == 0 )
+    {
+        LogDebug( ( "Poll timed out and there is no data to read from the buffer." ) );
+    }
+    else
+    {
+        LogError( ( "Poll returned with status = %d.", pollStatus ) );
+        bytesReceived = -1;
+    }
+
+    return bytesReceived;
+}
+
+/*-----------------------------------------------------------*/
+
+static uint32_t getTimeMs( void )
+{
+    uint32_t timeMs;
+    struct timespec timeSpec;
+
+    /* Get the MONOTONIC time. */
+    clock_gettime( CLOCK_MONOTONIC, &timeSpec );
+
+    /* Calculate the milliseconds from timespec. */
+    timeMs = ( uint32_t ) ( timeSpec.tv_sec * 1000 )
+             + ( uint32_t ) ( timeSpec.tv_nsec / ( 1000 * 1000 ) );
+
+    /* Reduce globalEntryTime from obtained time so as to always return the
+     * elapsed time in the application. */
+    timeMs = ( uint32_t ) ( timeMs - globalEntryTimeMs );
+
+    return timeMs;
+}
+
+/*-----------------------------------------------------------*/
+
+static int getNextFreeIndexForOutgoingPublishes( uint8_t * pIndex )
+{
+    int status = EXIT_FAILURE;
+
+    assert( outgoingPublishPackets != NULL );
+
+    for( *pIndex = 0; *pIndex < MAX_OUTGOING_PUBLISHES; ( *pIndex )++ )
+    {
+        /* A free index is marked by invalid packet id.
+         * Check if the the index has a free slot. */
+        if( outgoingPublishPackets[ *pIndex ].packetId == MQTT_PACKET_ID_INVALID )
+        {
+            status = EXIT_SUCCESS;
+            break;
+        }
+    }
+
+    return status;
+}
+/*-----------------------------------------------------------*/
+
+static void cleanupOutgoingPublishAt( uint8_t index )
+{
+    assert( outgoingPublishPackets != NULL );
+    assert( index < MAX_OUTGOING_PUBLISHES );
+
+    /* Clear the outgoing publish packet. */
+    ( void ) memset( &( outgoingPublishPackets[ index ] ),
```

Comment:
Nit: Since this demo uses static allocation and constant strings, this is correct, but should we have a comment reminding the user of that? If a user just copies our demo but instead uses dynamic allocation for payloads/topic names, they'd need to remember to free that memory pointed to by pPayload and pTopicName when cleaning up the publish.

Reply:
I think you have a good point to mention about the static allocation that we follow. I think instead of adding a caution statement for each of the allocation and free, it would be better to add a generic statement at the top.